### DISCUSSION TITLE
How do I create a Podspec for a command line tool?

### DISCUSSION THREAD

**Discussion by soconnor-florio** [SwiftGen/SwiftGen#828] [Q&A] [answered]
I didn't know where else to ask, so hopefully someone here might be able to help:

I'm creating a command line tool that's used for a build pipeline / run script on an iOS App.   We currently use SwiftGen in our app and installed it via Cocoapods, which is why I'm asking this group.

I'm curious to now how I can set up a command line tool that can be installed via Cocoapods, basically as SwiftGen has done.

I looked into the SwiftGen.podspec and I wasn't quite sure how a .zipfile as source translates to an executable in your Pods/SwiftGen folder.

Any help here, links to explanations, etc. would be very much appreciated.  I'm a huge fan of SwiftGen, excellent work!

**Comment by djbe** [comment, marked as answer]
CocoaPods doesn't support building the CLI tool on the fly for end-users. You're responsible for providing a binary that can directly be used.

What SwiftGen and many others (SwiftLint, Sourcery, ...) do is, for each release:
- Compile the tool into a binary + any related resources if needed (templates for SwiftGen)
- Zip these up
- Attach (upload) it to the GitHub release
- Point the podspec to that zip's URL, which will always be a certain pattern (github url, just change the version number to match your release's version number).

We do all these steps (and other things) using Rake, our steps for release can be found here:
https://github.com/SwiftGen/SwiftGen/blob/stable/rakelib/release.rake

Our process for a release is also documented here:
https://github.com/SwiftGen/SwiftGen/wiki/%5BAdmin%5D-Release-Process

**Reply by soconnor-florio** [reply]
That's terrific.  And thanks for predicting what I really wanted to use Cocoapods for; I thought it would build it on the fly, seeing as we will have to deal with Intel and Apple Silicon.  Is this a concern with pre-made binaries?

(update:  Ah, it seems I should read up on that a bit more.  #rosetta)

**Reply by djbe** [reply]
Well, ideally you'd avoid rosetta.

What you can do is create a ~"fat binary"~ "universal binary", which simply is a binary that contains a binary for each architecture. Using SPM this can be done with: `xcrun swift build -c release --arch arm64 --arch x86_64` for example. Read here for some more info: https://liamnichols.eu/2020/08/01/building-swift-packages-as-a-universal-binary.html

**Reply by djbe** [reply]
Couple of things to note:
- Our rake build has been simplified quite bit in our development branch `https://github.com/SwiftGen/SwiftGen/blob/develop/Rakefile` because we're switching over to SPM for managing our project (instead of Xcode)
- We actually still have to add those architecture flags, so good thing bringing up this topic 😅

**Reply by soconnor-florio** [reply]
Glad I could "help".  :D

I'd like to thank you again for taking the time to help me with this.  Really appreciate the effort.
